[24.1] Handle error when workflow is unowned in Invocation view

ahmedhamidawan opened the pull request:
Found this when investigating https://github.com/galaxyproject/galaxy/issues/18697; that if (in this case) I run another user's shared workflow with a subworkflow, the invocation view for the parent renders everything fine, but there is an unhandled error caused by the subworkflow not being shared. Now this fix handles that error and shows it on hover on the `WorkflowStepTitle`:

https://github.com/user-attachments/assets/7bf7c7e2-a8d9-4621-8652-d3a4b73df305

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
[24.1] Handle error when workflow is unowned in Invocation view

Found this when investigating https://github.com/galaxyproject/galaxy/issues/18697; that if (in this case) I run another user's shared workflow with a subworkflow, the invocation view for the parent renders everything fine, but there is an unhandled error caused by the subworkflow not being shared. Now this fix handles that error and shows it on hover on the `WorkflowStepTitle`

ahmedhamidawan committed:
prioritize link to subworkflow in invocation view

... and disable Report tab in subworkflow step

mvdbeek reviewed (approved):
More error handling is always great, thanks @ahmedhamidawan! We should also allow access to embedded workflows if we have access to the parent workflow.

ahmedhamidawan committed:
fix/add jest test for `WorkflowInvocationState`

ahmedhamidawan committed:
only show unique messages in `WorkflowInvocationOverview`

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationOverview.vue:
```diff
@@ -108,8 +108,10 @@ const invocationPdfLink = computed<string | null>(() => {
+const uniqueMessages = computed(() => {
+    const messages = props.invocation?.messages || [];
+    const uniqueMessagesSet = new Set(messages.map((message) => JSON.stringify(message)));
+    return Array.from(uniqueMessagesSet).map((message) => JSON.parse(message)) as InvocationMessageResponseModel[];
```
The manual type assignment here seems odd and likely to drift from the source defined in the API in the future. props.invocation.messages should already have the correct type, so type inference should also reveal the correct type on the set -> array conversion.

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationOverview.vue:
```diff
@@ -108,8 +108,10 @@ const invocationPdfLink = computed<string | null>(() => {
-    return props.invocation?.messages.length ? true : false;
+const uniqueMessages = computed(() => {
+    const messages = props.invocation?.messages || [];
+    const uniqueMessagesSet = new Set(messages.map((message) => JSON.stringify(message)));
```
Ah, why the stringify thing ?

mvdbeek replied:
I see, objects are not hashable. Well, that explains why you're assigning the type.

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationOverview.vue:
```diff
@@ -108,8 +108,10 @@ const invocationPdfLink = computed<string | null>(() => {
+const uniqueMessages = computed(() => {
+    const messages = props.invocation?.messages || [];
+    const uniqueMessagesSet = new Set(messages.map((message) => JSON.stringify(message)));
+    return Array.from(uniqueMessagesSet).map((message) => JSON.parse(message)) as InvocationMessageResponseModel[];
```
```suggestion
    const uniqueMessagesSet = new Set(messages.map((message) => JSON.stringify(message)));
    return Array.from(uniqueMessagesSet).map((message) => JSON.parse(message)) as typeof messages;
```

That would avoid any drift, assuming that stringify/parse doesn't change the type.

ahmedhamidawan replied:
![image](https://github.com/user-attachments/assets/00e92fb0-e9bb-418e-a6f9-72f8d024538c)

is the reason I assigned the `InvocationMessageResponseModel` type. And no, it is not stringify/parse changing the type, the error is being raised outside of this branch as well if you go to `WorkflowInvocationOverview`

I could also ignore it here since we were already doing so?

mvdbeek replied:
Have you tried my suggestion ? If this fails (it compiles fine for me) then I think there's something else wrong.

ahmedhamidawan replied:
That's what I tried in the screenshot. There might be something off on my end then 🤔 , I'll push your suggestion and you could confirm it compiles fine?

mvdbeek replied:
It compiles fine, otherwise the tests would fail. I do see this
<img width="1503" alt="Screenshot 2024-08-27 at 17 07 32" src="https://github.com/user-attachments/assets/e28e2569-7919-4532-bc18-fc219c3d60cd">

however. The solution to a problem like this can't be re-assigning the type ... this will never fail any compile-time checks, which we really have to avoid.

I think the proper fix is to drop the (semi-)hand-generated invocation interfaces in client/src/components/WorkflowInvocationState/invocationMessageModel.ts. These were only needed when we didn't have a fully-typed invocation model in the openapi docs, which we do now. I don't think there's actually a type-mismatch and it's a vetur issue, but of course it's cleaner to drop this.

I'll push a fix for that.

ahmedhamidawan replied:
That makes sense, thank you!

mvdbeek replied:
Turns out we already had some type drift: https://github.com/galaxyproject/galaxy/pull/18730/commits/e9eba0c620a3fb320b53b37a50403d31fa877763

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationOverview.vue:
```diff
@@ -3,10 +3,11 @@ import { BAlert, BButton } from "bootstrap-vue";
+import { type Workflow } from "@/stores/workflowStore";
 import { withPrefix } from "@/utils/redirect";
 
+import { type InvocationMessageResponseModel } from "./invocationMessageModel";
```
```suggestion
```

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue:
```diff
@@ -107,14 +107,17 @@ const invocationStateSuccess = computed(() => {
+const workflowVersion = computed(() => workflow.value?.version);
 
 onMounted(async () => {
     try {
```
I think you can pull all of this out of the onMounted hook ? Don't think we need to wait for the DOM to exist to start performing these calls.

ahmedhamidawan replied:
The reason I kept this all in the `onMounted` hook was because we had the 
```
await pollStepStatesUntilTerminal();
await pollJobStatesUntilTerminal();
```
calls in there that would start the timeout and keep trying to fetch even if there was an error raised from fetching the invocation. This way, we now only begin polling if we were able to fetch an invocation in the first place

mvdbeek replied:
you can do this in a watcher.

ahmedhamidawan replied:
Doing this and fetching the invocation (the first time; and catching any error) in a watcher

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationOverview.vue:
```diff
@@ -28,6 +29,7 @@ function getUrl(path: string): string {
 
 interface Props {
     invocation?: WorkflowInvocationElementView;
+    workflow?: Workflow;
```
I think it was cleaner, more performant and self-contained before ? `useWorkflowInstance` is great, it only fetches the workflow if we don't already have it, we now seem to always fetch the workflow even if it is already in store.

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue:
```diff
@@ -107,14 +107,17 @@ const invocationStateSuccess = computed(() => {
-        initialLoading.value = false;
+        invocationLoaded.value = true;
         if (invocation.value) {
+            await workflowStore.fetchWorkflowForInstanceId(invocation.value.workflow_id);
```
If all function calls in an async function are awaited, is it even async ? 

Also in this case `useWorkflowInstance` would be much preferable. If you want to improve the error handling there you could easily return a third object `error` in https://github.com/galaxyproject/galaxy/blob/dev/client/src/composables/useWorkflowInstance.ts#L16

ElectronicBlueberry replied:
> If all function calls in an async function are awaited, is it even async ?

Yes it is. The rest of the event loop continues to execute while the execution of this specific function is waiting.

In fact, the opposite is true:

If no function call in an async function is awaited, it is not really async.

mvdbeek replied:
> Yes it is. The rest of the event loop continues to execute while the execution of this specific function is waiting.

Correct me if I'm wrong, but the calls here will be executed linearly, just like in a sync function (yes, event loop continues) ? None of these calls however need to wait on each other, they're completely independent.

ElectronicBlueberry replied:
> Correct me if I'm wrong, but the calls here will be executed linearly, just like in a sync function

Which is the purpose of async functions. If no await is used, there is no need for the calling function to be async.

> None of these calls however need to wait on each other

the next line depends on this line.

ahmedhamidawan replied:
> the next line depends on this line

... is what I understood by using await calls as well, hence, I used them here.

mvdbeek replied:
This is the only line, and it can be replaced by `useWorkflowInstance `

mvdbeek replied:
So we're all on the same page that, if `useWorkflowInstance` is used, there's no need for this to be an async function, because nothing needs to be awaited ?

ahmedhamidawan replied:
The reason I didn't just place `useWorkflowInstance` outside of the `onMounted` hook and the if statement is because when this component is initialized, we have no invocation, and hence, no `invocation.workflow_id`, and so `useWorkflowInstance(invocation.value?.workflow_id)` causes an error (no workflow id).
Therefore, I added an `await fetchInvocationForId` and once that gives us the invocation, we fetch the workflow

I could replace that individual line:
```suggestion
            const { workflow: workflowIns } = useWorkflowInstance(invocation.value.workflow_id);
            workflow.value = workflowIns.value;
```

ahmedhamidawan replied:
I could also potentially set `workflow` as a `computed` ref like so:
```
const workflow = computed(() => {
    if (invocation.value) {
        const { workflow } = useWorkflowInstance(invocation.value.workflow_id);
        return workflow.value;
    }
    return undefined;
});
const isDeletedWorkflow = computed(() => workflow.value?.deleted === true);
const workflowVersion = computed(() => workflow.value?.version);

onMounted(async () => {
    try {
        await invocationStore.fetchInvocationForId({ id: props.invocationId });
        invocationLoaded.value = true;
        if (invocation.value) {
            await pollStepStatesUntilTerminal();
            await pollJobStatesUntilTerminal();
        }
    } catch (e) {
        errorMessage.value = errorMessageAsString(e);
    } finally {
        initialLoading.value = false;
    }
});
```

but the reason I had it the way it was in the singular mounted hook for fetching invocation AND workflow, was because we were catching errors from either there. That's why I changed `invocation` from a computed `getInvocationForId` to what we have now as well

mvdbeek replied:
I think it'd be a significant improvement to move this all out of the onMounted hook. The computed looks good to me (just need to return `error` in useWorkflowInstance).

ahmedhamidawan replied:
We are now not computing/getting the workflow in the parent `WorkflowInvocationState` component at all
Instead, we use it in a child `WorkflowInvocationHeader` and the existing `WorkflowInvocationOverview` where we use `useWorkflowInstance` which expects an `invocation.workflow_id` which we are ensuring exists in the parent.

ahmedhamidawan committed:
remove messages type from `WorkflowInvocationOverview`

Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>

mvdbeek committed:
Use canonical invocation message type from openapi schema

mvdbeek committed:
Create a named union for InvocationMessageResponseUnion

mvdbeek committed:
Fix incomplete type guard

ahmedhamidawan committed:
never fetch workflow in state component, fetch in child if !!invocation

Do not fetch workflow via explicit `workflowStore` functions or the `useWorkflowInstance` in the parent `WorkflowInvocationState` component, and instead place `useWorkflowInstance` in a new `WorkflowInvocationHeader` component which is only rendered if we have a valid invocation in the first place.

ahmedhamidawan committed:
fix grammar in invocation state hover text

ahmedhamidawan committed:
fix/adjust jests for errors shown in `WorkflowInvocationOverview`

mvdbeek reviewed (approved):
Awesome, thanks a lot for making all those changes!
